test/metabase/query_processor_test/cast_test.clj

```diff
@@ -213,6 +213,236 @@
+  (mt/test-drivers (mt/normal-drivers-with-feature :expressions/float)
+    (mt/dataset test-data
+      (let [mp (mt/metadata-provider)]
+        (doseq [[table expressions] [[:people [{:expression (lib/concat "1" ".5")
+                                                :db-type "TEXT"}
+                                               {:expression (lib/concat
+                                                             (lib.metadata/field mp (mt/id :people :id))
+                                                             "."
+                                                             (lib.metadata/field mp (mt/id :people :zip)))
+                                                :db-type "TEXT"}]]]
+                {:keys [expression db-type]} expressions]
+          (testing (str "Casting " db-type " to float")
+            (let [query (-> (lib/query mp (lib.metadata/table mp (mt/id table)))
+                            (lib/with-fields [(lib.metadata/field mp (mt/id table :id))])
+                            (lib/expression "UNCASTED" expression)
+                            (as-> q
+                                  (lib/expression q "FLOATCAST" (lib/float (lib/expression-ref q "UNCASTED"))))
+                            (lib/limit 10))
+                  result (-> query qp/process-query)
+                  cols (mt/cols result)
+                  rows (mt/rows result)]
+              (is (types/field-is-type? :type/Float (last cols)))
+              (doseq [[_ uncasted-value casted-value] rows]
+                (is (float= (double (Double/parseDouble uncasted-value))
+                            (double casted-value))
+                    (str "Text tested: " uncasted-value))))))))))
+
+(deftest ^:parallel float-cast-nested-native-query
+  (mt/test-drivers (mt/normal-drivers-with-feature :expressions/float)
+    (mt/dataset test-data
+      (let [mp (mt/metadata-provider)]
+        (doseq [{:keys [expression db-type]} [{:expression "'123.7'"  :db-type "TEXT"}
+                                              {:expression "'-123.1'" :db-type "TEXT"}]]
+          (testing (str "Casting " db-type " to float from native query")
+            (let [native-query (mt/native-query {:query (str "SELECT " expression " AS UNCASTED")})]
+              (mt/with-temp
+                [:model/Card
+                 {card-id :id}
+                 (mt/card-with-source-metadata-for-query native-query)]
+                (let [query (-> (lib/query mp (lib.metadata/card mp card-id))
+                                (lib/with-fields [])
+                                (as-> q
+                                      (lib/expression q "UNCAST" (->> q lib/visible-columns (filter #(= "uncasted" (u/lower-case-en (:name %)))) first)))
+                                (as-> q
+                                      (lib/expression q "FLOATCAST" (lib/float (->> q lib/visible-columns (filter #(= "uncasted" (u/lower-case-en (:name %)))) first)))))
```

Comment:
Wouldn't it make sense to extract `(->> q lib/visible-columns (filter #(= "uncasted" (u/lower-case-en (:name %)))) first)`? (I usually introduce a `base-query` in such situations, and then I don't need the ugly `as->`.)